include/boost/gil/extension/numeric/kernel.hpp

```diff
@@ -151,6 +153,203 @@ inline Kernel reverse_kernel(Kernel const& kernel)
     return result;
 }
 
+
+namespace detail {
+
+template <typename Core>
+class kernel_2d_adaptor : public Core
+{
+public:
+    kernel_2d_adaptor() = default;
+
+    explicit kernel_2d_adaptor(std::size_t center_vertical, std::size_t center_horizontal)
+        : center_(center_horizontal, center_vertical)
+    {
+        BOOST_ASSERT(center_.y < this->size() && center_.x < this->size());
+    }
+
+    kernel_2d_adaptor(std::size_t size, std::size_t center_vertical, std::size_t center_horizontal)
+        : Core(size * size),
+        square_size(size),
+        center_(center_horizontal, center_vertical)
+    {
+        BOOST_ASSERT(this->size() > 0);
+        BOOST_ASSERT(center_.y < this->size() && center_.x < this->size()); // also implies `size() > 0`
+    }
+
+    kernel_2d_adaptor(kernel_2d_adaptor const& other)
+        : Core(other),
+        square_size(other.square_size),
+        center_(other.center_.x, other.center_.y)
+    {
+        BOOST_ASSERT(this->size() > 0);
+        BOOST_ASSERT(center_.y < this->size() && center_.x < this->size()); // also implies `size() > 0`
+    }
+
+    kernel_2d_adaptor& operator=(kernel_2d_adaptor const& other)
+    {
+        Core::operator=(other);
+        center_.y = other.center_.y;
+        center_.x = other.center_.x;
+        square_size = other.square_size;
+        return *this;
+    }
+
+    std::size_t upper_size() const
+    {
+        BOOST_ASSERT(center_.y < this->size());
+        return center_.y;
+    }
+
+    std::size_t lower_size() const
+    {
+        BOOST_ASSERT(center_.y < this->size());
+        return this->size() - center_.y - 1;
+    }
+
+    std::size_t left_size() const
+    {
+        BOOST_ASSERT(center_.x < this->size());
+        return center_.x;
+    }
+
+    std::size_t right_size() const
+    {
+        BOOST_ASSERT(center_.x < this->size());
+        return this->size() - center_.x - 1;
+    }
+
+    auto center_vertical() -> std::size_t&
+    {
+        BOOST_ASSERT(center_.y < this->size());
+        return center_.y;
+    }
+
+    auto center_vertical() const -> std::size_t const&
+    {
+        BOOST_ASSERT(center_.y < this->size());
+        return center_.y;
+    }
+
+    auto center_horizontal() -> std::size_t&
+    {
+        BOOST_ASSERT(center_.x < this->size());
+        return center_.x;
+    }
+
+    auto center_horizontal() const -> std::size_t const&
+    {
+        BOOST_ASSERT(center_.x < this->size());
+        return center_.x;
+    }
+
+    std::size_t size() const
+    {
+        return square_size;
+    }
+
+    typename Core::value_type at(std::size_t x, std::size_t y) const
+    {
+        BOOST_ASSERT(0 <= x && x < size());
+        BOOST_ASSERT(0 <= y && y < size());
```

Comment:
@miralshah365 Instead of assertions, this should throw exceptional if x and y is put of range. See `std::vector::at` for reference and example

Reply:
done